src/main/java/org/jitsi/videobridge/cc/vp8/VP8AdaptiveTrackProjectionContext.java

```diff
@@ -0,0 +1,376 @@
+     *
+     * @param rtpPacket the VP8 packet to decide whether or not to project.
+     * @return true to project the packet, otherwise false.
+     */
+    private synchronized
+    VP8FrameProjection createVP8FrameProjection(
+        @NotNull RawPacket rtpPacket, int targetIndex)
+    {
+        // Creating a new VP8 projection depends on reading and results in
+        // writing of the last VP8 frame, therefore this method needs to be
+        // synced. At the time of this writing, only the translator thread is
+        // reaches this point.
+
+        VP8Frame lastVP8Frame = lastVP8FrameProjection.getVP8Frame();
+        // Old VP8 frames cannot be accepted because there's no "free" space in
+        // the sequence numbers. Check that before we create any structures to
+        // support the incoming packet/frame.
+        if (lastVP8Frame != null && lastVP8Frame.matchesOlderFrame(rtpPacket))
+        {
+            return null;
+        }
+
+        // if packet loss/re-ordering happened and this is not the first packet
+        // of a frame, then we don't process it right now. It'll get its chance
+        // when the first packet arrives and, if it's chosen for forwarding,
+        // we'll piggy-back any missed packets.
+        //
+        // This is to keep things simple (i.e. make it easy to compute the
+        // starting sequence number of the projection of an accepted frame).
+
+        byte[] buf = rtpPacket.getBuffer();
+        int payloadOff = rtpPacket.getPayloadOffset();
+        if (!DePacketizer.VP8PayloadDescriptor.isStartOfFrame(buf, payloadOff))
+        {
+            return null;
+        }
+
+        // Lastly, check whether the quality of the frame is something that we
+        // want to forward. We don't want to be allocating new objects unless
+        // we're interested in the quality of this frame.
+        if (!vp8QualityFilter.acceptFrame(rtpPacket, targetIndex))
+        {
+            return null;
+        }
+
+        long nowMs = System.currentTimeMillis();
+
+        // We know we want to forward this frame, but we need to make sure it's
+        // going to produce a decodable VP8 packet stream.
+        VP8FrameProjection nextVP8FrameProjection
+            = lastVP8FrameProjection.makeNext(rtpPacket, nowMs);
+        if (nextVP8FrameProjection == null)
+        {
+            return null;
+        }
+
+        // We have successfully projected the incoming frame and we've allocated
+        // a starting sequence number for it. Any previous frames can no longer
+        // grow.
+        vp8FrameProjectionMap.put(rtpPacket.getTimestamp(), nextVP8FrameProjection);
+        // The frame attached to the "last" projection is no longer the "last".
+        lastVP8FrameProjection = nextVP8FrameProjection;
+
+        // Cleanup the frame projection map.
+        vp8FrameProjectionMap.entrySet().removeIf(
+            e -> e.getValue().isFullyProjected(nowMs));
```

Comment:
is this only useful for cleaning the map?  if so and we want this based on time, maybe it should be called `isExpired`? which could imply it either 'timed out' or was completed successfully.

Reply:
Calling the method `isExpired` would describe what we want to do when the frame is fully projected and calling the method `isFullyProjected` would describe what the method checks. I can rename the method if you still think it's better (in light of this short explanation).

Reply:
I asked because I don't see `isFullyProjected` used anywhere but here (to clean the map), maybe I'm missing something in the diff?

Reply:
No you're not missing anything, the method is only used here. To me it's clearer the way it is because it makes it obvious that we cleanup the map because the frame is fully projected (i.e. I don't have to check the implementation). I don't really have a strong preference about the name and I'll rename it if you think it makes more sense.

Reply:
ok, i don't feel too strongly so can leave it as it is.